SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -39,180 +39,175 @@ void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
 	}
 }
 
-SpatialScopedActiveSpanId::SpatialScopedActiveSpanId(SpatialEventTracer* InEventTracer, const TOptional<Trace_SpanId>& InCurrentSpanId)
+SpatialScopedActiveSpanId::SpatialScopedActiveSpanId(SpatialEventTracer* InEventTracer, const FSpatialGDKSpanId& InCurrentSpanId)
 	: CurrentSpanId(InCurrentSpanId)
-	, EventTracer(InEventTracer->GetWorkerEventTracer())
+	, EventTracer(nullptr)
 {
-	if (InCurrentSpanId.IsSet())
+	if (InEventTracer == nullptr)
 	{
-		Trace_EventTracer_SetActiveSpanId(EventTracer, CurrentSpanId.GetValue());
+		return;
 	}
+
+	EventTracer = InEventTracer->GetWorkerEventTracer();
+	Trace_EventTracer_SetActiveSpanId(EventTracer, CurrentSpanId.GetConstId());
 }
 
 SpatialScopedActiveSpanId::~SpatialScopedActiveSpanId()
 {
-	if (CurrentSpanId.IsSet())
+	if (EventTracer != nullptr)
 	{
 		Trace_EventTracer_ClearActiveSpanId(EventTracer);
 	}
 }
 
 SpatialEventTracer::SpatialEventTracer(const FString& WorkerId)
 {
-	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>();
+	MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
+
+	Trace_EventTracer_Parameters parameters;
+	parameters.user_data = this;
+	parameters.callback = &SpatialEventTracer::TraceCallback;
+	EventTracer = Trace_EventTracer_Create(&parameters);
+
+	// Setup Event Tracing Sampling
+	Trace_SamplingParameters SamplingParameters;
+	SamplingParameters.sampling_mode = Trace_SamplingMode::TRACE_SAMPLING_MODE_PROBABILISTIC;
+
+	TArray<Trace_SpanSamplingProbability> SpanSamplingProbabilities;
+	TArray<TStringConversion<TStringConvert<TCHAR, ANSICHAR>, 128>> AnsiTypes;
+
+	for (const auto& Pair : Settings->EventSamplingModeOverrides)
 	{
-		if (Settings->bEventTracingEnabled)
-		{
-			MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
-			Enable(WorkerId);
-		}
+		int32 Index = AnsiTypes.Add(StringCast<ANSICHAR>(*Pair.Key.ToString()));
+		SpanSamplingProbabilities.Add({ AnsiTypes[Index].Get(), Pair.Value });
 	}
-}
 
-SpatialEventTracer::~SpatialEventTracer()
-{
-	if (IsEnabled())
+	SamplingParameters.probabilistic_parameters.default_probability = Settings->SamplingProbability;
+	SamplingParameters.probabilistic_parameters.probability_count = SpanSamplingProbabilities.Num();
+	SamplingParameters.probabilistic_parameters.probabilities = SpanSamplingProbabilities.GetData();
+
+	Trace_EventTracer_SetSampler(EventTracer, &SamplingParameters);
+
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing enabled."));
+
+	// Open a local file
+	FString EventTracePath = FPaths::Combine(FPaths::ProjectSavedDir(), TEXT("EventTracing"));
```

Comment:
Should this be a part of some constants file in-case it gets re-used someitme?

Reply:
Could be indeed! I think I will leave this for now. If we need it, it should be simple PR.